File: GOLANG/protocol/apicalls.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

const (
	AuthToken = "Team 17BC43C7-5CE5-4FBF-9AE5-1D3C1C66AA87"
	BaseURL   = "https://exs-htf-2024.azurewebsites.net"
)

func apiGet[T any](endpoint string) (T, error) {
	var result T
	client := &http.Client{}

	req, err := http.NewRequest("GET", BaseURL+endpoint, nil)
	if err != nil {
		return result, fmt.Errorf("failed to create request: %v", err)
	}

	req.Header.Add("Authorization", AuthToken)

	resp, err := client.Do(req)
	if err != nil {
		return result, fmt.Errorf("HTTP request error: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return result, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return result, fmt.Errorf("failed to read response body: %v", err)
	}

	err = json.Unmarshal(body, &result)
	if err != nil {
		return result, fmt.Errorf("JSON deserialization error: %v", err)
	}

	return result, nil
}

func apiPost[T any](endpoint string, body T) error {
	client := &http.Client{}

	jsonBody, err := json.Marshal(body)
	if err != nil {
		return fmt.Errorf("failed to serialize body: %v", err)
	}

	req, err := http.NewRequest("POST", BaseURL+endpoint, bytes.NewBuffer(jsonBody))
	if err != nil {
		return fmt.Errorf("failed to create request: %v", err)
	}

	req.Header.Add("Authorization", AuthToken)
	req.Header.Add("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("HTTP request error: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	return nil
}

File: Shared/APICalls.cs
using System.Text.Json;
using System.Text;

namespace Shared
{
    public static class APICalls
    {
        private static readonly string AUTH_TOKEN = "Team 17BC43C7-5CE5-4FBF-9AE5-1D3C1C66AA87";
        private static readonly string BASE_URL = "https://exs-htf-2024.azurewebsites.net";

        public static async Task<T> Get<T>(string endpoint)
        {

            using (var client = new HttpClient())
            {
                // Add custom auth header
                client.DefaultRequestHeaders.Add("Authorization", AUTH_TOKEN);

                try
                {
                    HttpResponseMessage response = await client.GetAsync($"{BASE_URL}{endpoint}");

                    // Ensure successful status code`
                    response.EnsureSuccessStatusCode();

                    // Read and deserialize the response content
                    string jsonResponse = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<T>(jsonResponse);
                }
                catch (HttpRequestException ex)
                {
                    Console.WriteLine($"HTTP Request Error: {ex.Message}");
                }
                catch (JsonException ex)
                {
                    Console.WriteLine($"JSON Deserialization Error: {ex.Message}");
                }
            }

            Console.WriteLine("Failed to get");
            System.Environment.Exit(0);
            return default(T);
        }

        public static async Task Post<T>(string endpoint, T body)
        {
            using (var client = new HttpClient())
            {
                // Add custom auth header
                client.DefaultRequestHeaders.Add("Authorization", AUTH_TOKEN);

                try
                {
                    string jsonBody = JsonSerializer.Serialize(body);

                    var content = new StringContent(jsonBody, Encoding.UTF8, "application/json");

                    var response = await client.PostAsync($"{BASE_URL}{endpoint}", content);

                    // Ensure successful status code
                    response.EnsureSuccessStatusCode();
                }
                catch (HttpRequestException ex)
                {
                    Console.WriteLine($"HTTP Request Error: {ex.Message}");
                }
                catch (JsonException ex)
                {
                    Console.WriteLine($"JSON Deserialization Error: {ex.Message}");
                }
            }
        }
    }
}
